### PR TITLE
Add functionality to enable skipping analyzers for implicitly trigger…

### PR DESCRIPTION
**Pull request by mavasani**
…ed builds inside Visual Studio

Work towards [AB#1337109](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1337109)

## Overview

For builds which are implicitly triggered inside Visual Studio from commands such as 'Run Tests' or 'Start Debugging', we implicitly skip analyzers to speed up these builds.

Such builds will set a special property `IsImplicitlyTriggeredBuild` to `true`, which guards the implicit skip analyzers logic. We display a special message to inform the users about us implicitly skipping analyzers to speed up the build. Note this message was provided to us by the UX team.

To enable MSBuild's incremental build logic and project system's fast-upto-date check logic to work correctly, we create/touch a special semaphore file to indicate the time stamp for last build with skipAnalyzers flag. This semaphore file is passed as a custom additional file input to builds without skipAnalyzers flag to ensure correct incremental builds. Additionally, we pass this file as an 'UpToDateCheckInput' item for project system's fast-upto-date check.

## Testing

[AB#1337109](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1337109) requires few more WIP PRs to go through to work end-to-end:
1. Changes on Unit testing repo (Run tests and Debug tests commands) and Start Debugging command to set `IsImplicitlyTriggeredBuild` to `true` for builds indirectly triggered from these commands. I am working on these changes.
2. Change to project-system repo to update the fast-upto-date check logic to work correctly. @drewnoakes is driving that work.

I was able to locally build bits with both the above changes + changes from this PR and verify end-to-end functioning of this feature for Run Tests/Debug Tests command. Demo provided below.

## Demo

ImplicitlySkipAnalyzersForIndirectBuilds

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
> IsIndirectlyTriggeredBuildInsideVisualStudi

How about:

IsImplicitlyTriggeredBuild?  I'm not sure why we'd need to say it was visual studio, and I imagine this being useful for other hosts.

File: src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets [hunk 1]
```diff
@@ -70,6 +77,42 @@
...
+
+    <!-- PERF: For builds which are indirectly triggered inside Visual Studio from commands 
+               such as from 'Run Tests' or 'Start Debugging', we implicitly skip analyzers to speed up the build.
+               We only do so by default when 'TreatWarningsAsErrors' is not 'true'. For the scenario where
+			         'TreatWarningsAsErrors' is 'true', users can explicitly enable this functionality by setting
+			         'OptimizeIndirectlyTriggeredBuildInsideVisualStudio' to 'true'.
+    -->
+    <PropertyGroup Condition="'$(_SkipAnalyzers)' == '' and
+                              '$(IsIndirectlyTriggeredBuildInsideVisualStudio)' == 'true' and
+                              ('$(TreatWarningsAsErrors)' != 'true' or '$(OptimizeIndirectlyTriggeredBuildInsideVisualStudio)' == 'true')">
```
Comment (jaredpar) [inline src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets:89]: Consider that editor config can also cause warnings to be promoted to errors. That is another hidden way that warnings can be promoted. Should we be suppressing additional files when we are skipping analyzers? They shouldn't really have a purpose in that mode correct?

File: src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets [hunk 2]
```diff
@@ -70,6 +77,42 @@
...
+      <_SkipAnalyzers>true</_SkipAnalyzers>
+    </PropertyGroup>
+
+    <!-- Display a message to inform the users about us implicitly skipping analyzers for speeding up indirect builds. -->
+    <Message Condition="'$(_ImplicitlySkipAnalyzers)' == 'true'"
+             Text="Skipping analyzers to speed up the build. You can execute 'Build' or 'Rebuild' command to run analyzers." Importance="High" />
+
+    <!-- Semaphore file to indicate the time stamp for last build with skipAnalyzers flag. -->
+    <PropertyGroup>
+      <LastBuildWithSkipAnalyzers>$(IntermediateOutputPath)$(MSBuildProjectFile).BuildWithSkipAnalyzers</LastBuildWithSkipAnalyzers>
```
Comment (jaredpar) [inline src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets:100]: Isn't this property local to this file? If so we should give it the `_` prefix.

Comment (jaredpar) [inline src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets:100]: Have we verified that this logic works correctly with a multi-targeted build? Essentially does it create one of these files per target framework? Such that each TF is indpendently evaluated? 

I can't remember off hand if `$(IntermediateOutputPath)` is TF specific or not at this point in the build

**Comment by mavasani** [comment]
> IsImplicitlyTriggeredBuild? I'm not sure why we'd need to say it was visual studio, and I imagine this being useful for other hosts.

That sounds reasonable, I'll update the naming.

File: src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets (hunk 1 above)
Comment (mavasani) [reply src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets:89]: > Consider that editor config can also cause warnings to be promoted to errors. That is another hidden way that warnings can be promoted.

Yes, that is a known design decision as we lose majority of perf gains associated with the feature if we end up loading analyzers and then identifying if any of them report an error in each file and conditionally enable/disable them. We believe majority of users would retain analyzer diagnostics as warnings for their local development with warnaserror in CI to catch them. Additionally, amongst such users percentage of users who would oppose us unblocking debugging/running tests in presence of such analyzer errors would also be pretty small, if any. So for v1 we are going with the current design of only disabling the feature by default for TreatWarningsAsErrors, and iterate on it based on customer feedback.

> Should we be suppressing additional files when we are skipping analyzers? They shouldn't really have a purpose in that mode correct?

Possibly, but I am unsure if that can have side effects. For example, code fixes for compiler warnings can potentially rely on additional files.

File: src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets (hunk 2 above)
Comment (mavasani) [reply src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets:100]: Thanks. Each TFM gets a separate intermediate path: `<%project_dir%>\obj\<%config%>\<%tfm%>\`. I just tested it works fine for multi-TFM project.

Comment (mavasani) [reply src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets]: Thanks, updated.

**Review by jaredpar — APPROVED** [review]

**Comment by jaredpar** [comment]
@chsienki, @RikkiGibson for second review

File: src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets (hunk 1 above)
Comment (chsienki) [reply src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets]: >Should we be suppressing additional files when we are skipping analyzers? They shouldn't really have a purpose in that mode correct?

They are still used by generators and so required for correctness of the build.

**Review by chsienki — APPROVED** [review]
Same as @jared said: as long as we're explicitly aware that a user can get potentially different compilation outcomes between test + build, LGTM.

File: src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets (hunk 1 above)
Comment (jaredpar) [reply src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets]: 🤦‍♂️ yep, feel silly for not thinking of that.

**Pull request merged**